src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,223 @@
+
+    if (source->hasFormat("image/png"))
+    {
+        uploadImageToNuuls({source->data("image/png"), "png"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/jpeg"))
+    {
+        uploadImageToNuuls({source->data("image/jpeg"), "jpeg"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/gif"))
+    {
+        uploadImageToNuuls({source->data("image/gif"), "gif"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasUrls())
+    {
+        // This path gets chosen when files are copied from a file manager, like explorer.exe, caja.
+        // Each entry in source->urls() is a QUrl pointing to a file that was copied.
+        for (const QUrl &path : source->urls())
+        {
+            if (!getImageFileFormat(path.toLocalFile()).isEmpty())
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading image: %1").arg(path.toLocalFile())));
+                QImage img = QImage(path.toLocalFile());
+                if (img.isNull())
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Couldn't load image :(")));
+                    return;
+                }
+
+                boost::optional<QByteArray> imageData = convertToPng(img);
+                if (imageData)
+                {
+                    TypedBytes data = {imageData.get(), "png"};
+                    uploadQueue.push(data);
+                }
+                else
+                {
+                    channel->addMessage(makeSystemMessage(
+                        QString("Cannot upload file: %1, Couldn't convert "
+                                "image to png.")
+                            .arg(path.toLocalFile())));
+                }
+            }
+            else if (path.toLocalFile().endsWith(".gif"))
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Uploading GIF: %1").arg(path.toLocalFile())));
+                QFile file(path.toLocalFile());
+                bool isOkay = file.open(QIODevice::ReadOnly);
+                if (!isOkay)
+                {
+                    channel->addMessage(
+                        makeSystemMessage(QString("Failed to open file. :(")));
+                    return;
+                }
+                TypedBytes data = {file.readAll(), "gif"};
+                uploadQueue.push(data);
+                file.close();
+                // file.readAll() => might be a bit big but it /should/ work
+            }
+            else
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Cannot upload file: %1, not an image")
+                        .arg(path.toLocalFile())));
+                isUploading = false;
+            }
+        }
```

Comment:
path.toLocalFile() is called too many times here - put it in a variable and use that